### ISSUE TITLE
Model support to RKNN conversion

### ISSUE THREAD

**Issue by Miguel-S-Marques** [ankandrew/fast-plate-ocr#46]
I am trying to convert the global plates model to RKNN but 2 ONNX OPs are not supported ("Neg" and "Reciprocal").
Do you know any backbone that meets all the supported ONNX OPs mentioned below?

| **Operators**             | **Remarks**                             |
| ------------------------- | --------------------------------------- |
| Abs                       | Not Supported                           |
| Acos                      | Not Supported                           |
| Acosh                     | Not Supported                           |
| Add                       |                                         |
| And                       |                                         |
| ArgMax                    |                                         |
| ArgMin                    |                                         |
| Asin                      | Not Supported                           |
| Asinh                     | Not Supported                           |
| Atan                      | Not Supported                           |
| Atanh                     | Not Supported                           |
| AveragePool               |                                         |
| BatchNormalization        |                                         |
| Bernoulli                 | Not Supported                           |
| BitShift                  | Not Supported                           |
| BitwiseAnd                | Not Supported                           |
| BitwiseNot                | Not Supported                           |
| BitwiseOr                 | Not Supported                           |
| BitwiseXor                | Not Supported                           |
| BlackmanWindow            | Not Supported                           |
| Cast                      |                                         |
| CastLike                  | Not Supported                           |
| Ceil                      | Not Supported                           |
| Celu                      | Not Supported                           |
| CenterCropPad             | Not Supported                           |
| Clip                      |                                         |
| Col2Im                    | Not Supported                           |
| Compress                  | Not Supported                           |
| Concat                    |                                         |
| ConcatFromSequence        | Not Supported                           |
| Constant                  |                                         |
| ConstantOfShape           |                                         |
| Conv                      |                                         |
| ConvInteger               | Not Supported                           |
| ConvTranspose             |                                         |
| Cos                       |                                         |
| Cosh                      | Not Supported                           |
| CumSum                    | Not Supported                           |
| DeformConv                | Not Supported                           |
| DepthToSpace              |                                         |
| DequantizeLinear          |                                         |
| Det                       | Not Supported                           |
| DFT                       | Not Supported                           |
| Div                       |                                         |
| Dropout                   |                                         |
| DynamicQuantizeLinear     | Not Supported                           |
| Einsum                    | Not Supported                           |
| Elu                       |                                         |
| Equal                     |                                         |
| Erf                       |                                         |
| Exp                       |                                         |
| Expand                    |                                         |
| EyeLike                   | only support constant input             |
| Flatten                   |                                         |
| Floor                     |                                         |
| Gather                    |                                         |
| GatherElements            |                                         |
| GatherND                  | Not Supported                           |
| Gemm                      |                                         |
| GlobalAveragePool         |                                         |
| GlobalLpPool              | Not Supported                           |
| GlobalMaxPool             |                                         |
| Greater                   |                                         |
| GreaterOrEqual            |                                         |
| GridSample                | Not Supported                           |
| GroupNormalization        | Not Supported                           |
| GRU                       | batchsize: 1                            |
| HammingWindow             | Not Supported                           |
| HannWindow                | Not Supported                           |
| Hardmax                   | Not Supported                           |
| HardSigmoid               |                                         |
| HardSwish                 |                                         |
| Identity                  |                                         |
| If                        | only support constant input             |
| InstanceNormalization     |                                         |
| IsInf                     | Not Supported                           |
| IsNaN                     | Not Supported                           |
| LayerNormalization        |                                         |
| LeakyRelu                 |                                         |
| Less                      |                                         |
| LessOrEqual               |                                         |
| Log                       |                                         |
| LogSoftmax                | batchsize: 1                            |
| Loop                      | Not Supported                           |
| LpNormalization           |                                         |
| LpPool                    | Not Supported                           |
| LRN                       |                                         |
| LSTM                      |                                         |
| MatMul                    |                                         |
| MatMulInteger             | Not Supported                           |
| Max                       |                                         |
| MaxPool                   |                                         |
| MaxRoiPool                |                                         |
| MaxUnpool                 |                                         |
| Mean                      | Not Supported                           |
| MeanVarianceNormalization |                                         |
| MelWeightMatrix           | Not Supported                           |
| Min                       |                                         |
| Mish                      |                                         |
| Mod                       |                                         |
| Mul                       |                                         |
| Multinomial               | Not Supported                           |
| Neg                       | Not Supported                           |
| NegativeLogLikelihoodLoss | Not Supported                           |
| NonMaxSuppression         | Not Supported                           |
| NonZero                   | Not Supported                           |
| Not                       | Not Supported                           |
| OneHot                    | Not Supported                           |
| Optional                  | Not Supported                           |
| OptionalGetElement        | Not Supported                           |
| OptionalHasElement        | Not Supported                           |
| Or                        | Not Supported                           |
| Pad                       |                                         |
| Pow                       |                                         |
| PRelu                     |                                         |
| QLinearConv               | Not Supported                           |
| QLinearMatMul             | Not Supported                           |
| QuantizeLinear            |                                         |
| RandomNormal              | Not Supported                           |
| RandomNormalLike          | Not Supported                           |
| RandomUniform             | Not Supported                           |
| RandomUniformLike         | Not Supported                           |
| Range                     | Not Supported                           |
| Reciprocal                | Not Supported                           |
| ReduceL1                  | Not Supported                           |
| ReduceL2                  | Not Supported                           |
| ReduceLogSum              | Not Supported                           |
| ReduceLogSumExp           | Not Supported                           |
| ReduceMax                 |                                         |
| ReduceMean                |                                         |
| ReduceMin                 |                                         |
| ReduceProd                | Not Supported                           |
| ReduceSum                 |                                         |
| ReduceSumSquare           | Not Supported                           |
| Relu                      |                                         |
| Reshape                   |                                         |
| Resize                    | mode: nearest2d/bilinear                |
| ReverseSequence           |                                         |
| RNN                       | Not Supported                           |
| RoiAlign                  | pool type: average<br />batchsize: 1    |
| Round                     | Not Supported                           |
| Scan                      | Not Supported                           |
| ScatterElements           | Not Supported                           |
| ScatterND                 |                                         |
| Selu                      | Not Supported                           |
| SequenceAt                | Not Supported                           |
| SequenceConstruct         | Not Supported                           |
| SequenceEmpty             | Not Supported                           |
| SequenceErase             | Not Supported                           |
| SequenceInsert            | Not Supported                           |
| SequenceLength            | Not Supported                           |
| SequenceMap               | Not Supported                           |
| Shape                     |                                         |
| Shrink                    | Not Supported                           |
| Sigmoid                   |                                         |
| Sign                      | Not Supported                           |
| Sin                       |                                         |
| Sinh                      | Not Supported                           |
| Size                      |                                         |
| Slice                     | batchsize: 1                            |
| Softmax                   | batchsize: 1                            |
| SoftmaxCrossEntropyLoss   | Not Supported                           |
| Softplus                  |                                         |
| Softsign                  | Not Supported                           |
| SpaceToDepth              |                                         |
| Split                     |                                         |
| SplitToSequence           | Not Supported                           |
| Sqrt                      |                                         |
| Squeeze                   |                                         |
| STFT                      | Not Supported                           |
| StringNormalizer          | Not Supported                           |
| Sub                       |                                         |
| Sum                       | Not Supported                           |
| Tan                       | Not Supported                           |
| Tanh                      |                                         |
| TfIdfVectorizer           | Not Supported                           |
| ThresholdedRelu           | Not Supported                           |
| Tile                      | batchsize: 1<br />not support broadcast |
| TopK                      | Not Supported                           |
| Transpose                 |                                         |
| Trilu                     | Not Supported                           |
| Unique                    | Not Supported                           |
| Unsqueeze                 |                                         |
| Where                     |                                         |
| Xor                       | Not Supported                           |

**Comment by ankandrew** [comment]
Seems like the `Neg` and  operator is caused by the LayerNorm. I would try `ConvMixer` which should be a good alternative for current models and uses BN instead of LN which doesn't seem to export with these 2 ops. I can try DyT from new [paper](https://arxiv.org/abs/2503.10622) from Meta (replacement for LN), which shouldn't use these 2 problematic ops.

**Comment by ankandrew** [comment]
You can replace the backbone in the code if you really want. Bare in mind new version of this `fast-plate-ocr` is coming and I will experiment using DyT to avoid this bad operators.

**Comment by Miguel-S-Marques** [comment]
Can you send me a trained model with the new architecture with DyT to check if I can convert it without problems?

**Comment by ankandrew** [comment]
Btw, this model expects the following input shape: `[N, 64, 128, 3]`.

**Comment by Miguel-S-Marques** [comment]
I forgot to mention that the model needs to have the following input shape:  [1, 3, 64, 128]

NCHW format with batch size 1.

**Comment by ankandrew** [comment]
Fixed batch size with `NHWC`, for channel first I have to make more code changes, can't do it now:

[dyt-model-fixed-batch.onnx.zip](https://github.com/user-attachments/files/20505044/dyt-model-fixed-batch.onnx.zip)

I don't see the op you are mentioning in the graph, the `ReduceProd`. Can you inspect it with `Netron` and pinpoint where that operator is?

**Comment by Miguel-S-Marques** [comment]
And I am not sure but I think that the input and output need to be float32 to be able to convert to RKNN

**Comment by ankandrew** [comment]
Probably you need to change the following files:

- https://github.com/ankandrew/fast-plate-ocr/blob/ankandrew/new-default-backbone/fast_plate_ocr/train/model/layers.py
- https://github.com/ankandrew/fast-plate-ocr/blob/ankandrew/new-default-backbone/fast_plate_ocr/train/model/models.py

You can also try to mimic current architecture by using DyT as Norm, a couple of conv and MultiHead (nothing else is actually used besides that) and try exporting that ro RKNN.

**Comment by Miguel-S-Marques** [comment]
I was trying to modify the code but I couldn't export it.

**Comment by ankandrew** [comment]
Since this is for doing inference outside `fast-plate-ocr` and seems like a 1-time export, I prefer to just keep things as they are and leave an example in the documentation. I'll update the export function to make it work with channel first/last and support different dtype. After that this is just as simple as wrapping the trained model (would work with any model given current input signature):

```py
from pathlib import Path

import keras
import numpy as np

from fast_plate_ocr.cli.export import export_onnx
from fast_plate_ocr.train.model.config import load_plate_config_from_yaml
from fast_plate_ocr.train.model.model_builders import build_model
from fast_plate_ocr.train.model.model_schema import load_model_config_from_yaml

PLATE_CONFIG_DIR = Path(__file__).resolve().parent / "config"
MODELS_CONFIG_DIR = Path(__file__).resolve().parent / "models"

plate_config_file = PLATE_CONFIG_DIR / "default_latin_plate_config.yaml"
plate_config = load_plate_config_from_yaml(plate_config_file)
model_config = load_model_config_from_yaml(MODELS_CONFIG_DIR / "cct_s_v1.yaml")
existing_model = build_model(model_config, plate_config)
existing_model(np.ones((1, 64, 128, 3), dtype=np.uint8))
existing_model.summary(expand_nested=True)

inp_nchw = keras.Input(shape=(3, 64, 128), dtype="float32", name="input_nchw")
to_nhwc = keras.layers.Permute([2, 3, 1])  # N C H W -> N H W C
out = existing_model(to_nhwc(inp_nchw))
model_nchw = keras.Model(inputs=inp_nchw, outputs=out, name="plate_ocr_nchw")
export_onnx(  # Assuming corrected version, dtype and channel mode will be passed here
    model=model_nchw,
    plate_config=plate_config,
    out_file=Path("model_nchw.onnx"),
    simplify=True,
    dynamic_batch=False,
    skip_validation=False,
)
```

**Comment by ankandrew** [comment]
Hey @Miguel-S-Marques I've added support for input float32 dtype and channels first. Keep in mind to fully remove the `ReduceProd` operator you need to use `dynamic_batch=False` and `simplify=True`. Besides that with the new version you can do:

```
--onnx-input-dtype,
float32,
--onnx-data-format,
channels_first,
```

so it's suitable for RKNN. I've added tests around this too.
